File: D.S.A/arrays/medium/greedy/Buy-sell-stock-2.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    
    let profit=0;
    let min=prices[0];

     for(let i=0;i<prices.length;i++){
      if(prices[i]<min){
        min=prices[i];
      }
      if(prices[i]-min>0){
        profit+=prices[i]-min;
        min=prices[i];
      }

     }

     return profit;
};
// Explanation:
// - Track current buying price (`min`).
// - If today is cheaper, update buying price.
// - If profitable, add gain and treat today as new buy.
// - This captures all increasing segments.
// - Return total profit from multiple trades.


// Complexity:
// - Time: O(n) because prices are scanned once and every rise is added.
// - Space: O(1) because only a few tracking variables are used.

File: Java/dsa/arrays/medium/greedy/BuySellStock2.java
public class BuySellStock2 {
    public static int maxProfit(int[] prices) {
        int profit = 0;
        int min = prices[0];
        for (int price : prices) {
            if (price < min) {
                min = price;
            }
            if (price - min > 0) {
                profit += price - min;
                min = price;
            }
        }
        return profit;
    }
}
